### PR TITLE
Fixes concurrent map write iteration issue

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/5376
Towards https://github.com/giantswarm/giantswarm/issues/5414

We have a case where the map of labels to histograms that the `histogramvec` type wraps is written to in the `Add` and `Ensure` method, while the map itself can be iterated over (as we just return the map in the `Histograms` method).

This PR introduces a lock into the `Histograms` method, and creates a copy of the internal map which is then returned. This means that the internal map is never iterated over (as the copy is, instead).

tl;dr - without the fix, the test blows up with a `fatal error: concurrent map iteration and map write`, and doesn't with it :D

### CODE REVIEW

File: histogramvec/histogramvec.go [hunk 1]
```diff
@@ -87,7 +87,16 @@ func (hv *HistogramVec) Ensure(labels []string) {
...
+// Histograms returns a copy of the currently managed Histograms.
 func (hv *HistogramVec) Histograms() map[string]*histogram.Histogram {
-	return hv.histograms
+	hv.mutex.Lock()
+	defer hv.mutex.Unlock()
+
+	histogramCopy := map[string]*histogram.Histogram{}
+
+	for label, histogram := range hv.histograms {
+		histogramCopy[label] = histogram
```
Comment (TheoBrigitte) [inline histogramvec/histogramvec.go]: Here you only get a different map holding the same pointers if I'm not mistaken. If you want to make a full copy here you have to get a new pointer for each `histogram`.

```go
h:=*histogram
histogramCopy[label] = &h
```

Comment (JosephSalisbury) [reply histogramvec/histogramvec.go]: That does right to me.

Would the new pointer just point to the original struct though? I feel like I need to copy the entire struct.

**Comment by JosephSalisbury** [comment]
@TheoBrigitte ptala

File: histogramvec/histogramvec.go (hunk 1 above)
Comment (TheoBrigitte) [reply histogramvec/histogramvec.go]: The new pointer (`&h`) would point to the copied value (`h`) and no longer the original one (`*histogram`). So no need for further copy.

File: histogramvec/histogramvec.go [hunk 2]
```diff
@@ -87,7 +87,17 @@ func (hv *HistogramVec) Ensure(labels []string) {
 	}
 }
 
-// Histograms returns the currently managed Histograms.
+// Histograms returns a copy of the currently managed Histograms.
 func (hv *HistogramVec) Histograms() map[string]*histogram.Histogram {
```
Comment (tomahawk28) [inline histogramvec/histogramvec.go]: Isn't that a problem we can solve it by value receiver?

Comment (TheoBrigitte) [reply histogramvec/histogramvec.go]: No, even if we would use value receiver, we would still have to copy the pointer within the histogram.
